File: Chapter11/Exercise11_06/Exercise11_06b.scala
import org.apache.spark.sql.types.{IntegerType, StringType, StructField, StructType}
import org.apache.spark.sql.{Row, SparkSession}

object Exercise11_06b {

  def main(args: Array[String]): Unit = {

    // Build a SparkSession in Local Mode
    val spark = SparkSession
      .builder()
      .master("local[2]")
      .appName("My Spark App")
      .getOrCreate()

    import spark.implicits._
    val my_previous_pets = Seq(Row("annabelle", "cat"),
                               Row("daisy", "kitten"),
                               Row("roger", "puppy"),
                               Row("joe", "puppy dog"),
                               Row("rosco", "dog"),
                               Row("julie", "feline"))

    val schema = List(
      StructField("nickname", StringType, nullable = true),
      StructField("type", StringType, nullable = true)
    )

    val petsRDD = spark.sparkContext.parallelize(my_previous_pets)
    val petsDF = spark.createDataFrame(petsRDD, StructType(schema))

    case class Pet(nickname: String, petType: String)

    val standardized_pets = petsDF.map(pet => {
      val nickname = pet.getString(pet.fieldIndex("nickname"))
      val petType = pet.getString(pet.fieldIndex("type"))

      println(nickname, petType)
      val standardType =
        if (Seq("dog", "puppy", "puppy dog", "hound", "canine").contains(petType)){
          "dog"
        }
        else if (Seq("cat", "kitty", "kitten", "feline", "kitty cat").contains(petType)){
        "cat"
        }
      else{
          petType
        }

      (nickname, standardType)
    })

    standardized_pets.show()

  }

}

File: Chapter11/Exercise11_06/Exercise11_06b.py
from pyspark.sql import SparkSession
from pyspark.sql import Row

# Create a Spark Session
spark = SparkSession \
    .builder \
    .appName("My Spark App") \
    .master("local[2]") \
    .getOrCreate()

# create spark contexts
sc = spark.sparkContext

my_previous_pets = [Row("annabelle", "cat"),
                    Row("daisy", "kitten"),
                    Row("roger", "puppy"),
                    Row("joe", "puppy dog"),
                    Row("rosco", "dog"),
                    Row("julie", "feline")]

# create RDDs
petsRDD = sc.parallelize(my_previous_pets)

def standardize(pet):

    name = pet[0]
    animal_type = pet[1]

    if animal_type in ["dog", "puppy", "puppy dog", "hound", "canine"]:
        return Row(name, "dog")
    elif animal_type in ["cat", "kitty", "kitten", "feline", "kitty cat"]:
        return Row(name, "cat")
    else:
        return pet


standardizedPets = petsRDD.map(standardize)
standardizedPetsDF = spark.createDataFrame(standardizedPets, ['nickname', 'type'])
standardizedPetsDF.show()
